qemu/tests/thin_provisioning.py

```diff
@@ -0,0 +1,169 @@
+    provisioning_mode = get_provisioning_mode(host_id, disk_name)
+    logging.info("Scsi pci id and plot is: {:s} and disk name is: {:s}".format(host_id, disk_name))
+    logging.info("Current provisioning_mode = {:s}", provisioning_mode)
+
+    # Bitmap must be empty before test
+    if get_allocation_bitmap():
+        logging.debug("Block allocation bitmap: %s" % get_allocation_bitmap())
+        raise error.TestError("Block allocation bitmap not empty before test.")
+
+    vm_name = params["main_vm"]
+    test_image = params.get("test_image")
+    params["start_vm"] = "yes"
+    params["image_name_%s" % test_image] = disk_name
+    params["image_format_%s" % test_image] = "raw"
+    params["image_raw_device_%s" % test_image] = "yes"
+    params["force_create_image_%s" % test_image] = "no"
+    params["drv_extra_params_scsi_debug"] = "discard=on"
+    params["images"] = " ".join([params["images"], test_image])
+
+    error.context("TEST STEP 1: Boot guest with disk {:s}".format(disk_name), logging.info)
+    env_process.preprocess_vm(test, params, env, vm_name)
+    vm = env.get_vm(vm_name)
+    vm.verify_alive()
+    timeout = float(params.get("login_timeout", 240))
+    session = vm.wait_for_login(timeout=timeout)
+
+    guest_device_name = get_guest_discard_disk(session)
+    rewrite_disk_cmd = params["rewrite_disk_cmd"]
+    rewrite_disk_cmd = rewrite_disk_cmd.replace("DISK", guest_device_name)
+    error.context("TEST STEP 2: Create a file in {:s}, using the command {:s}".format(guest_device_name,
+                                                                                      rewrite_disk_cmd), logging.info)
+    session.cmd(rewrite_disk_cmd, timeout=timeout, ignore_all_errors=True)
+    bitmap_before_trim = get_allocation_bitmap()
+    error.context("TEST STEP 3: Bitmap before test: {:s}".format(bitmap_before_trim), logging.info)
+    if not re.match(r"\d+-\d+", bitmap_before_trim):
+        raise exceptions.TestFail("Bitmap should be continuous before fstrim")
+
+    format_disk_cmd = params["format_disk_cmd"]
+    format_disk_cmd = format_disk_cmd.replace("DISK1", guest_device_name)
+    error.context("TEST STEP 3: Format disk '{:s}' in guest".format(guest_device_name), logging.info)
+    session.cmd(format_disk_cmd)
+
+    error.context("TEST STEP 4: Mount disk with discard options {:s}".format(guest_device_name),
+                  logging.info)
+    mount_disk_cmd = params["mount_disk_cmd"]
+    mount_disk_cmd = mount_disk_cmd.replace("DISK1", guest_device_name)
+    session.cmd(mount_disk_cmd)
+
+    error.context("TEST STEP 5: Execute fstrim in guest", logging.info)
+    fstrim_cmd = params["fstrim_cmd"]
+    o = session.cmd_output(fstrim_cmd, timeout=timeout)
+    if params.get("drive_format_{:s}".format(test_image), None) == "scsi-hd":
+        pass
+    else:
+        if not re.search(params["match_str"], o):
+            exceptions.TestError("Commond output is wrong,please check.", logging.error)
```

Comment:
Refer (can save several lines which is more clear in the code ):
if not params.get("drive_format_{:s}".format(test_image), None) == "scsi-hd":
    if not re.search(params["match_str"], o):
        exceptions.TestError("Commond output is wrong,please check.", logging.error)